src/utils/nostr.tsx

```diff
@@ -0,0 +1,153 @@
+/* eslint-disable react-refresh/only-export-components */
+import NDK, { NDKNip07Signer, NDKUser } from "@nostr-dev-kit/ndk";
+import { createContext, useContext, useEffect, useState } from "react";
+
+export const RELAYS = ["wss://relay.snort.social", "wss://nos.lol"];
+
+export interface NostrContextResult {
+  /**
+   * A connected instance of the Nostr Dev Kit class
+   */
+  ndk: NDK | undefined;
+  /**
+   * An NDKUser instance representing the current user over `window.nostr`
+   */
+  user: NDKUser | undefined;
+  /**
+   * Whether the Nostr connection is loading
+   */
+  isLoading: boolean;
+  /**
+   * If an error occurred in attempt to connect to Nostr.
+   */
+  error: Error | null;
+}
+
+interface NostrPending extends NostrContextResult {
+  ndk: undefined;
+  user: undefined;
+  isLoading: true;
+  error: null;
+}
+
+interface NostrErrorResult extends NostrContextResult {
+  ndk: undefined;
+  user: undefined;
+  isLoading: false;
+  error: Error;
+}
+
+interface NostrSuccessResult extends NostrContextResult {
+  ndk: NDK;
+  user: NDKUser;
+  isLoading: false;
+  error: null;
+}
+
+type NostrProviderType = NostrPending | NostrErrorResult | NostrSuccessResult;
+
+export const NostrContext = createContext<NostrProviderType | null>(null);
+
+/**
+ * Connects to `window.nostr`, initializing and exposing `user` and `ndk` through `NostrConnectionContext`.
+ */
+export function NostrProvider({ children }: { children: React.ReactNode }) {
+  const [data, setData] = useState<{ user: NDKUser; ndk: NDK } | undefined>(undefined);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<Error | null>(null);
+
+  useEffect(() => {
+    async function init() {
+      try {
+        const signer = new NDKNip07Signer();
+        const ndk = new NDK({
+          explicitRelayUrls: RELAYS,
+          signer,
+        });
+
+        await ndk.connect(2500);
+
+        if (
+          "nostr" in window &&
+          window.nostr &&
+          (("isEnabled" in window.nostr &&
+            typeof window.nostr.isEnabled === "function" &&
+            (await window.nostr?.isEnabled())) ||
+            ("isEnabled" in window.nostr &&
+              typeof window.nostr.isEnabled === "boolean" &&
+              window.nostr.isEnabled) ||
+            ("_isEnabled" in window.nostr && window.nostr?._isEnabled))
+        ) {
```

Comment:
you don't need all these typechecks, just define the window.nostr global interface somewhere and assert that window.nostr exists